### ISSUE TITLE
Memory error with timsort and std::deque

### ISSUE THREAD

**Issue by Morwenn** [Morwenn/cpp-sort#89]
Valgrind sometimes reports memory errors when trying to sort an `std::deque` with `tim_sorter` (see the deque-memcheck branch for more information).

It probably comes from out-of-bounds `std::deque` iterators. Since they are never dereferenced, it never actually causes any error, and Valgrind doesn't complain about equivalent out-of-bounds pointers. I don't know which kind of problems these out-of-bound `std::deque` iterators may cause, but we should probably better get rid of them as soon as possible.

**Comment by Morwenn** [comment]
May go away if I switch the implementation to tvanslyke/timsort-cpp

**Comment by Morwenn** [comment]
I have added new tests for `std::deque` to `develop` but haven't managed to trigger a memory error yet.

**Comment by Morwenn** [comment]
I might have fixed the issue with the latest commit: when maintaining timsort upstream I got to test it with MSVC debug iterators and they detected that some iterators were going out-of-bounds. Apparently it was due to `cursor1` in `mergeHi` sometimes pointing to the element right before the beginning of a collection, and even though said element was never dereferenced not every collection can safely form such an iterator (while it isn't an issue with contiguous iterators, I'm not surprised that it caused issues with `std::deque`). Most of the Valgrind traces point to `mergeHi`, which is quite consistent with the issue found by the MSVC debug iterators.

I fixed the issue by tweaking the way iterators are handled in `mergeHi`, but the only failing build in cpp-sort (OSX with Valgrind) now fails because it lasts longer than 50 minutes, so I can't really close this issue for now. I'll close it when I have enough non-failing builds again to be sure.
